controllers/storageclaim_controller.go

```diff
@@ -501,64 +539,111 @@ func (r *StorageClaimReconciler) getCephRBDStorageClass(data map[string]string)
 		},
-		Driver:         csi.GetRBDDriverName(),
+		Driver:         driverName,
 		DeletionPolicy: snapapi.VolumeSnapshotContentDelete,
 		Parameters:     data,
 	}
-	return volumesnapshotclass
+	return volumegroupsnapshotclass
 }
 
-func (r *StorageClaimReconciler) createOrReplaceStorageClass(storageClass *storagev1.StorageClass) error {
-	existing := &storagev1.StorageClass{}
-	existing.Name = r.storageClaim.Name
+func (r *StorageClaimReconciler) createOrReplaceClass(class client.Object) error {
+	var existing client.Object
+	var existingParameters map[string]string
+	var classParameters map[string]string
+	classKind := class.GetObjectKind().GroupVersionKind().Kind
+	switch classKind {
+	case volumeSnapshotClassKind:
+		existing = &snapapi.VolumeSnapshotClass{}
+		classParameters = class.(*snapapi.VolumeSnapshotClass).Parameters
+	case volumeGroupSnapshotClassKind:
+		existing = &groupsnapapi.VolumeGroupSnapshotClass{}
+		classParameters = class.(*groupsnapapi.VolumeGroupSnapshotClass).Parameters
+	case "StorageClass":
+		existing = &storagev1.StorageClass{}
+		classParameters = class.(*storagev1.StorageClass).Parameters
+	default:
+		return fmt.Errorf("failed. %s class type unsupported", classKind)
+	}
 
-	if err := r.own(storageClass); err != nil {
-		return fmt.Errorf("failed to own storageclass: %v", err)
+	existing.SetName(r.storageClaim.Name)
+
+	if err := r.own(class); err != nil {
+		return fmt.Errorf("failed to own %s: %v", classKind, err)
 	}
 
 	if err := r.get(existing); err != nil && !errors.IsNotFound(err) {
-		return fmt.Errorf("failed to get StorageClass: %v", err)
+		return fmt.Errorf("failed to get %s: %v", classKind, err)
 	}
 
-	// If present then compare the existing StorageClass with the received StorageClass, and only proceed if they differ.
-	if reflect.DeepEqual(existing.Parameters, storageClass.Parameters) {
-		return nil
+	switch classKind {
+	case volumeSnapshotClassKind:
+		existingParameters = existing.(*snapapi.VolumeSnapshotClass).Parameters
+	case volumeGroupSnapshotClassKind:
+		existingParameters = existing.(*groupsnapapi.VolumeGroupSnapshotClass).Parameters
+	case "StorageClass":
+		existingParameters = existing.(*storagev1.StorageClass).Parameters
+	default:
+		return fmt.Errorf("failed. %s class type unsupported", classKind)
 	}
 
-	// StorageClass already exists, but parameters have changed. Delete the existing StorageClass and create a new one.
-	if existing.UID != "" {
+	// If present then compare the existing Storage(Volume*Snapshot)Class parameters and labels with
+	// the received Storage(Volume*Snapshot)Class parameters and labels, and only proceed if they differ.
+	labelsEqual := reflect.DeepEqual(existing.GetLabels(), class.GetLabels())
+	parametersEqual := reflect.DeepEqual(existingParameters, classParameters)
+	annotationsEqual := reflect.DeepEqual(existing.GetAnnotations(), class.GetAnnotations())
+	if labelsEqual && parametersEqual && annotationsEqual {
+		return nil
+	}
```

Comment:
You're describing the implementation of `CreateOrUpdate`, yet what is required is to Recreate (delete and create) in case of inequality. Also, it's not exactly an ever-growing list per se as these would be the only sections of the CR that need comparing.

Reply:
> You're describing the implementation of CreateOrUpdate

Yes, think about it as a `recreateInChanged` that accepts a resource and a mutate function. It internally creates a copy of the given resource, runs the mutate function on top of it, and then compares the original object with the new mutated one.  if they differ it calls Delete on the resource and then Create on the same resource with the mutated payload.